x/incentive/keeper/keeper.go

```diff
@@ -981,3 +981,89 @@ func (k Keeper) GetAllClaims(ctx sdk.Context) types.Claims {
 
 	return cs
 }
+
+// GetRewardAccrualTime fetches the last time rewards were accrued for the
+// specified ClaimType and subKey.
+func (k Keeper) GetRewardAccrualTime(
+	ctx sdk.Context,
+	claimType types.ClaimType,
+	subKey string,
+) (time.Time, bool) {
+	store := prefix.NewStore(ctx.KVStore(k.key), types.GetPreviousRewardAccrualTimeKeyPrefix(claimType))
+	b := store.Get([]byte(subKey))
+	if b == nil {
+		return time.Time{}, false
+	}
+	var accrualTime types.AccrualTime
+	k.cdc.MustUnmarshal(b, &accrualTime)
+
+	return accrualTime.PreviousAccumulationTime, true
+}
+
+// SetRewardAccrualTime stores the last time rewards were accrued for the
+// specified ClaimType and subKey.
+func (k Keeper) SetRewardAccrualTime(
+	ctx sdk.Context,
+	claimType types.ClaimType,
+	subKey string,
+	blockTime time.Time,
+) {
+	store := prefix.NewStore(ctx.KVStore(k.key), types.GetPreviousRewardAccrualTimeKeyPrefix(claimType))
+
+	at := types.NewAccrualTime(claimType, subKey, blockTime)
+	bz := k.cdc.MustMarshal(&at)
+	store.Set([]byte(subKey), bz)
+}
+
+// IterateRewardAccrualTimes iterates over all reward accrual times of a given
+// claimType and performs a callback function.
+func (k Keeper) IterateRewardAccrualTimes(
+	ctx sdk.Context,
+	claimType types.ClaimType,
+	cb func(string, time.Time) (stop bool),
+) {
+	store := prefix.NewStore(ctx.KVStore(k.key), types.GetPreviousRewardAccrualTimeKeyPrefix(claimType))
+	iterator := sdk.KVStorePrefixIterator(store, []byte{})
+	defer iterator.Close()
+	for ; iterator.Valid(); iterator.Next() {
+		var accrualTime types.AccrualTime
+		k.cdc.MustUnmarshal(iterator.Value(), &accrualTime)
+
+		if cb(accrualTime.CollateralType, accrualTime.PreviousAccumulationTime) {
+			break
+		}
+	}
+}
+
+// IterateAllRewardAccrualTimes iterates over all reward accrual times of any
+// claimType and performs a callback function.
+func (k Keeper) IterateAllRewardAccrualTimes(
```

Comment:
naming suggestion: It can be clearer to have the `IterateX` method iterate over all `X` and `IterateXByY` methods for iterating over `X` in groups. Makes it easier to add additional Ys, like `IterateClaimsByOwner`, `IterateClaimsByClaimType` (not that that applies for `AccrualTimes`)

Reply:
good idea for later iteration groups